pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -522,63 +501,129 @@ func (p *Pod) constructGroupPodSets(podsInGroup corev1.PodList) ([]kueue.PodSet,
+	if err != nil {
+		return err
+	}
+	originalQueue := jobframework.QueueName(p)
 
-	var (
-		podSets []kueue.PodSet
-	)
+	if len(activePods) < groupTotalCount {
+		errMsg := fmt.Sprintf("'%s' group total count is less than the actual number of pods in the cluster", p.groupName())
+		r.Eventf(p.Object(), corev1.EventTypeWarning, "ErrWorkloadCompose", errMsg)
+		return jobframework.UnretryableError(errMsg)
+	}
 
-	if !p.isGroup {
-		podSets = p.PodSets()
-	} else {
-		groupTotalCount, err := p.groupTotalCount()
-		if err != nil {
-			return nil, err
+	for _, podInGroup := range p.list.Items {
+		// Skip failed pods
+		if podInGroup.Status.Phase == corev1.PodFailed {
+			continue
 		}
 
-		podCount := 0
-		for i := range p.list.Items {
-			if p.list.Items[i].Status.Phase != corev1.PodFailed {
-				podCount++
-			}
+		if podInGroupQueue := jobframework.QueueNameForObject(&podInGroup); podInGroupQueue != originalQueue {
+			return jobframework.UnretryableError(fmt.Sprintf("pods '%s' and '%s' has different queue names: %s!=%s",
+				p.pod.GetName(), podInGroup.GetName(),
+				originalQueue, podInGroupQueue))
 		}
 
-		if podCount != groupTotalCount {
-			errMsg := fmt.Sprintf("'%s' group total count is different from the actual number of pods in the cluster", p.groupName())
-			r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", errMsg)
-			return nil, jobframework.UnretryableError(errMsg)
+		tc, err := strconv.Atoi(podInGroup.GetAnnotations()[GroupTotalCountAnnotation])
+		if err != nil {
+			return fmt.Errorf("failed to extract '%s' annotation from the pod '%s': %w",
+				GroupTotalCountAnnotation,
+				podInGroup.GetName(),
+				err)
 		}
+		if tc != groupTotalCount {
+			return jobframework.UnretryableError(fmt.Sprintf("pods '%s' and '%s' has different '%s' values: %d!=%d",
+				p.pod.GetName(), podInGroup.GetName(),
+				GroupTotalCountAnnotation,
+				groupTotalCount, tc))
+		}
+	}
 
-		podSets, err = p.constructGroupPodSets(p.list)
-		if err != nil {
-			if jobframework.IsUnretryableError(err) {
-				r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", err.Error())
+	return nil
+}
+
+// activePods returns a slice of non-failed pods in the group
+func (p *Pod) activePods() []corev1.Pod {
+	// Filter all failed pods
+	activePods := make([]corev1.Pod, 0, len(p.list.Items))
+	for _, pod := range p.list.Items {
+		if pod.Status.Phase != corev1.PodFailed {
+			activePods = append(activePods, pod)
+		}
+	}
+
+	return activePods
+}
+
+// cleanupExcessPods will delete and finalize pods created last if the number of
+// activePods is greater than the totalCount value.
+func (p *Pod) cleanupExcessPods(ctx context.Context, c client.Client, totalCount int, activePods []corev1.Pod) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	extraPodsCount := len(activePods) - totalCount
+
+	if extraPodsCount <= 0 {
+		return nil
+	}
+
+	// Sort active pods by creation timestamp
+	sort.SliceStable(activePods, func(i, j int) bool {
```

Comment:
do we need it to be stable? Regular sort uses less memory.

Reply:
You're right. If there's two pods with equal creation timestamp we don't care which one to delete. 

Updated to unstable sort in 35fad1d1